src/couch_replicator/src/couch_replicator_auth.erl

```diff
@@ -0,0 +1,100 @@
+% Licensed under the Apache License, Version 2.0 (the "License"); you may not
+% use this file except in compliance with the License. You may obtain a copy of
+% the License at
+%
+%   http://www.apache.org/licenses/LICENSE-2.0
+%
+% Unless required by applicable law or agreed to in writing, software
+% distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+% WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+% License for the specific language governing permissions and limitations under
+% the License.
+
+-module(couch_replicator_auth).
+
+
+-export([
+    initialize/1,
+    update_headers/2,
+    handle_response/4,
+    cleanup/1
+]).
+
+
+-include_lib("couch_replicator/include/couch_replicator_api_wrap.hrl").
+
+
+-type headers() :: [{string(), string()}].
+-type code() :: non_neg_integer().
+
+
+-define(DEFAULT_PLUGINS,
+    "couch_replicator_auth_session,couch_replicator_auth_basic").
+
+
+% Behavior API
+
+-callback initialize(#httpdb{}) -> {ok, #httpdb{}, term()} | ignore.
+
+-callback update_headers(term(), headers()) -> {headers(), term()}.
```

Comment:
this makes the (reasonable) assumption that all auth systems work via headers, but it isn't an ideal abstraction. The intention behind allowing a callback to update headers is to apply the authentication whatevers, so perhaps that would be a better shape?

Reply:
It was a practical decision by looking at some of authentications mechanisms. Most I could find seem to work with headers. OAuth2 add the bearer token as a header. Session cookie is a header as well.

Other things I could think of are the url itself, so it allows adding query params, and the body. I don't know if passing the body through it would be a good idea (thinking of large streaming attachments for example). And couldn't think of a practical use for updating the url.